lib/datadog/di/probe_notification_builder.rb

```diff
@@ -0,0 +1,216 @@
+      end
+
+      def build_snapshot(probe, rv: nil, snapshot: nil,
+        duration: nil, callers: nil, args: nil, kwargs: nil,
+        serialized_entry_args: nil)
+        # TODO also verify that non-capturing probe does not pass
+        # snapshot or vars/args into this method
+        captures = if probe.capture_snapshot?
+          if probe.method?
+            {
+              entry: {
+                # standard:disable all
+                arguments: if serialized_entry_args
+                  serialized_entry_args
+                else
+                  (args || kwargs) && serializer.serialize_args(args, kwargs)
+                end,
+                throwable: nil,
+                # standard:enable all
+              },
+              return: {
+                arguments: {
+                  "@return": serializer.serialize_value(rv),
+                },
+                throwable: nil,
+              },
+            }
+          elsif probe.line?
+            {
+              lines: snapshot && {
+                probe.line_no => {locals: serializer.serialize_vars(snapshot)},
+              },
+            }
+          end
+        end
+
+        location = if probe.line?
+          actual_file = if probe.file
+            # Normally callers should always be filled for a line probe
+            # but in the test suite we don't always provide all arguments.
+            callers&.detect do |caller|
+              File.basename(caller.sub(/:.*/, '')) == File.basename(probe.file)
+            end&.sub(/:.*/, '') || probe.file
+          end
+          {
+            file: actual_file,
+            lines: [probe.line_no],
+          }
+        elsif probe.method?
+          {
+            method: probe.method_name,
+            type: probe.type_name,
+          }
+        end
+
+        stack = if callers
+          format_callers(callers)
+        end
+
+        timestamp = timestamp_now
+        {
+          service: settings.service,
+          "debugger.snapshot": {
+            id: SecureRandom.uuid,
+            timestamp: timestamp,
+            evaluationErrors: [],
+            probe: {
+              id: probe.id,
+              version: 0,
+              location: location,
+            },
+            language: 'ruby',
+            # TODO add test coverage for callers being nil
+            stack: stack,
+            captures: captures,
+          },
+          # In python tracer duration is under debugger.snapshot,
+          # but UI appears to expect it here at top level.
+          duration: duration ? (duration * 10**9).to_i : nil,
+          host: nil,
+          logger: {
+            name: probe.file,
+            method: probe.method_name || 'no_method',
+            thread_name: Thread.current.name,
+            thread_id: thread_id,
+            version: 2,
+          },
+          "dd.trace_id": 136035165280417366521542318182735500431,
+          "dd.span_id": 17576285113343575026,
```

Comment:
These look like placeholders -- should there be a FIXME or something here?

Reply:
Good catch, I forgot about these completely - they probably need to be filled out somehow.

Reply:
But, I  can do that in a separate PR.